src/app/services/auth/auth.service.ts

```diff
@@ -1,26 +1,152 @@
+  }
 
-@Injectable()
-export class AuthService implements CanActivate {
   constructor(
-    private router: Router,
-    private ws: WebSocketService,
-    @Inject(WINDOW) private window: Window,
+    private wsManager: WebsocketConnectionService,
   ) { }
 
-  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): boolean {
-    if (this.ws.loggedIn) {
-      return true;
-    }
+  login(username: string, password: string, otp: string = null): Observable<boolean> {
+    const uuid = UUID.UUID();
+    this.wsManager.send({
+      id: uuid,
+      msg: IncomingApiMessageType.Method,
+      method: 'auth.login',
+      params: otp ? [username, password, otp] : [username, password],
+    });
+    return this.getFilteredWebsocketResponse<boolean>(uuid).pipe(tap((response) => {
+      this.isLoggedIn = response;
+      this.isLoggedIn$.next(response);
+    }));
+  }
+
+  loginWithToken(): Observable<boolean> {
+    const uuid = UUID.UUID();
+    this.wsManager.send({
+      id: uuid,
+      msg: IncomingApiMessageType.Method,
+      method: 'auth.login_with_token',
+      params: [this.token2],
+    });
+    return this.getFilteredWebsocketResponse<boolean>(uuid).pipe(tap((response) => {
+      this.isLoggedIn = response;
+      this.isLoggedIn$.next(response);
+    }));
+  }
+
+  getFilteredWebsocketResponse<T>(uuid: string): Observable<T> {
+    return this.wsManager.websocketSubject$.pipe(
+      filter((data: IncomingWebsocketMessage) => data.msg === IncomingApiMessageType.Result && data.id === uuid),
+      map((data: ResultMessage<T>) => data.result),
+      take(1),
+    );
+  }
+
+  generateToken(tokenLifetime: number): Observable<string> {
+    const uuid = UUID.UUID();
+    const payload: {
+      id: string;
+      msg: IncomingApiMessageType;
+      method: string;
+      params?: [number];
+    } = {
+      id: uuid,
+      msg: IncomingApiMessageType.Method,
+      method: 'auth.generate_token',
+      params: [tokenLifetime],
+    };
+    this.wsManager.send(payload);
+    return this.getFilteredWebsocketResponse<string>(uuid);
+  }
+
+  generateTokenWithDefaultLifetime(): Observable<string> {
+    const uuid = UUID.UUID();
+    const payload = {
+      id: uuid,
+      msg: IncomingApiMessageType.Method,
+      method: 'auth.generate_token',
+    };
+    this.wsManager.send(payload);
+    return this.getFilteredWebsocketResponse<string>(uuid);
+  }
+
+  logout(): Observable<void> {
+    const uuid = UUID.UUID();
+    const payload = {
+      id: uuid,
+      msg: IncomingApiMessageType.Method,
+      method: 'auth.logout',
+    };
+    this.wsManager.send(payload);
+    return this.getFilteredWebsocketResponse<void>(uuid).pipe(tap(() => {
+      this.isLoggedIn = false;
+      this.isLoggedIn$.next(false);
+    }));
+  }
 
-    this.window.sessionStorage.setItem('redirectUrl', state.url);
-    this.router.navigate(['/sessions/signin']);
+  private getLoggedInUserInformation(): void {
```

Comment:
Wrong line. This method is unused and it makes `Change Password` in user menu hidden.